private/pkg/storage/storagearchive/storagearchive.go

```diff
@@ -91,25 +103,51 @@ func Untar(
 		if err := walkChecker.Check(ctx); err != nil {
 			return err
 		}
+		if tarHeader.Size < 0 {
+			return fmt.Errorf("invalid size for tar file %s: %d", tarHeader.Name, tarHeader.Size)
+		}
+		var fileReader io.Reader = tarReader
+		if options.maxFileSize > 0 {
+			fileReader = io.LimitReader(fileReader, options.maxFileSize)
+		}
 		path, ok, err := unmapArchivePath(tarHeader.Name, mapper, stripComponentCount)
 		if err != nil {
 			return err
 		}
-		if !ok {
-			continue
-		}
-		if tarHeader.FileInfo().Mode().IsRegular() {
-			if tarHeader.Size < 0 {
-				return fmt.Errorf("invalid size for tar file %s: %d", tarHeader.Name, tarHeader.Size)
+		if !ok || !tarHeader.FileInfo().Mode().IsRegular() {
+			// Next also discards the file but doesn't apply the limit.
+			// Explicitly discard the file with limit applied.
```

Comment:
We can apply a file limit without using `io.LimitReader`.

```go
if tarHeader.Size > options.maxFileSize {
    return ErrFileSizeLimit
}
```

No special casing requires if the default `options.maxFileSize` value is `math.MaxInt64`.